distribution/xfer/transfer.go

```diff
@@ -155,18 +154,23 @@ func (t *transfer) Watch(progressOutput progress.Output) *Watcher {
 		defer func() {
 			close(w.running)
 		}()
-		done := false
+		var (
+			done           bool
+			lastWritten    progress.Progress
+			hasLastWritten bool
+		)
 		for {
 			t.mu.Lock()
 			hasLastProgress := t.hasLastProgress
 			lastProgress := t.lastProgress
 			t.mu.Unlock()
 
-			// This might write the last progress item a
-			// second time (since channel closure also gets
-			// us here), but that's fine.
-			if hasLastProgress {
+			// Make sure we don't write the last progress item
+			// twice.
+			if hasLastProgress && (!done || !hasLastWritten || lastProgress != lastWritten) {
```

Comment:
Yes, that would be possible. But the goal here is to avoid a double-send of the last line when the watcher is detached. I wasn't aiming to filter out messages that are explicitly sent twice. That said, maybe there's no valid use case for sending repeated messages, and filtering them out in all cases would be a slight simplification. No strong feelings on this.